File: src/cpp/examples/mpiring/LHC.java
/*
* Copyright 2008, 2009 Complex Automata Simulation Technique (COAST) consortium
* Copyright 2010-2013 Multiscale Applications on European e-Infrastructures (MAPPER) project
*
* GNU Lesser General Public License
* 
* This file is part of MUSCLE (Multiscale Coupling Library and Environment).
* 
* MUSCLE is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* MUSCLE is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
* 
* You should have received a copy of the GNU Lesser General Public License
* along with MUSCLE.  If not, see <http://www.gnu.org/licenses/>.
*/
package examples.mpiring;

import com.sun.jna.Callback;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.ptr.DoubleByReference;
import muscle.core.ConduitExit;
import muscle.core.CxADescription;

public class LHC extends muscle.core.kernel.CAController {
	
	private ConduitExit<double[]> pipeInput;

  	public interface MPIRING extends Library {
		public void Ring_Init(String ringName);
		public void Ring_Broadcast_Params(DoubleByReference deltaE, DoubleByReference maxE);
		public double insertProton(double initialEnergy, double energyThreshold, EnergyCallback energyCallback);
		public boolean  isMasterNode();
		public void accelerateProtons( double deltaE, double energyThreshold);
		public void Ring_Cleanup();
   	}

	public  interface EnergyCallback extends Callback {
    		void newEnergy(double energy);
  	}
	
	static EnergyCallback energyCallback = new EnergyCallback(){
        	public void newEnergy(double energy) {
             		System.out.println("LHC: Proton energy callback: " + energy);
            		}
		};

	static MPIRING ring;

	static {
		ring = (MPIRING) Native.loadLibrary("mpiring", MPIRING.class);
	}

	public void addPortals() {
		pipeInput = addExit("pipe", double[].class);
	}

	protected void execute() {
	
		ring.Ring_Init("LHC");

        if (ring.isMasterNode()) {
        	double deltaEnergy = CxADescription.ONLY.getDoubleProperty("LHC:DeltaEnergy");
        	double maxEnergy = CxADescription.ONLY.getDoubleProperty("LHC:MaxEnergy");
        	
        	ring.Ring_Broadcast_Params(new DoubleByReference(deltaEnergy), new DoubleByReference(maxEnergy));
        	
        	double initialEnergy = pipeInput.receive()[0];
        	System.err.println("LHC: Received proton from PSB: " +  initialEnergy);
        	
        	double finalEnergy = ring.insertProton(initialEnergy, maxEnergy, energyCallback);
        	System.err.println("LHC: Final energy: " +  finalEnergy);
        } else {
        	DoubleByReference deltaEnergy = new DoubleByReference();
        	DoubleByReference maxEnergy = new DoubleByReference();
        	
        	ring.Ring_Broadcast_Params(deltaEnergy, maxEnergy);

            ring.accelerateProtons(deltaEnergy.getValue(), maxEnergy.getValue());
        }

        ring.Ring_Cleanup();
	}	

}

File: src/cpp/examples/mpiring/LHC.c
/*
* Copyright 2010-2013 Multiscale Applications on European e-Infrastructures (MAPPER) project
*
* GNU Lesser General Public License
* 
* This file is part of MUSCLE (Multiscale Coupling Library and Environment).
* 
* MUSCLE is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* MUSCLE is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
* 
* You should have received a copy of the GNU Lesser General Public License
* along with MUSCLE.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <stdio.h>
#include <stdlib.h>

#include <muscle2/cmuscle.h>

#include "mpiringlib.h"

static void energy_callback(double energy)
{
        static int loop = 0;

        printf("LHC:Energy in loop %d: %f\n", ++loop, energy);
}


int main(int argc, char **argv)
{

	Ring_Init("LHC");

	if (isMasterNode())
	{
		double energy = - 1, deltaEnergy = -1, maxEnergy = -1;
		size_t count = 1;
		int will_stop = 0;

		MUSCLE_Init(&argc, &argv); /* MUSCLE calls are only permitted in the rank 0 process */

		will_stop = MUSCLE_Will_Stop();
		deltaEnergy = atof(MUSCLE_Get_Property("LHC:DeltaEnergy"));
		maxEnergy = atof(MUSCLE_Get_Property("LHC:MaxEnergy"));
		
		while (!will_stop) {
			/* wrapper over MPI_Bcast */
			Ring_Broadcast_Params(&deltaEnergy, &maxEnergy, &will_stop);

			MUSCLE_Receive("pipe-in", &energy, &count, MUSCLE_DOUBLE);

			printf("LHC:Received proton energy: %f\n", energy);

			energy = insertProton(energy, maxEnergy, energy_callback);
			
			will_stop = MUSCLE_Will_Stop();
		}
		Ring_Broadcast_Params(&deltaEnergy, &maxEnergy, &will_stop);

		printf("LHC:Final proton energy: %f\n", energy);

		MUSCLE_Finalize();

	} else
	{
		double deltaEnergy = -1, maxEnergy = -1;
		int will_stop = 0;

		while (1) {
			/* wrapper over MPI_Bcast */
			Ring_Broadcast_Params(&deltaEnergy, &maxEnergy, &will_stop);
			if (will_stop) break;
			
			accelerateProtons(deltaEnergy, maxEnergy);
		}
	}

	Ring_Cleanup();
	return 0;

}
